### ISSUE TITLE
Add rule to exclude a path, like the inverse of `pick`

### ISSUE THREAD

**Issue by MatthewEppelsheimer** [buildinspace/peru#150]
Feature request: In cases where there are lots of directories and files to include from a dependency module _except one_, it would be nice to be able to specify explicit paths to omit completely, rather than using `pick` to verbosely declare everything to include.

I would suggest either:

1. Add an `omit` or `exclude` keyword for use in rule definitions
2. Support exclusion rules as part of `pick` statements with bangs, e.g. `pick: ! hey/neat/this/could/also/support/globbing/**`

I think I prefer 2.

**Comment by oconnor663** [comment]
I've been thinking about adding a `drop` field for a long time. I think @olson-sean-k was too. Name bikeshedding starts..._now!_ :)

I think I prefer option 1, because then we won't need to figure out a story for escaping file names that begin with `!`. (We do already support escaping `*`'s, and I just pushed 724dd92d0361d1a6c21fcb49996cc126b91b80d4 because I realized we weren't testing it.)

**Comment by MatthewEppelsheimer** [comment]
Bikeshedding woo! In no particular order, with a range of seriousness:

- `drop`
- `exclude`
- `omit`
- `kick`
- `forget`
- `ignore`
- `excommunicate`
- `disbar`
- `abandon`
- `disavow`
- `ban`
- `annul`
- `excoriate`

**Comment by oconnor663** [comment]
```yaml
purge: foo.txt
disappear: foo.txt
nuke: foo.txt
```

**Comment by oconnor663** [comment]
You know actually I never considered `rm`/`delete` either :p

**Comment by MatthewEppelsheimer** [comment]
Although my personal impractical favorite is `excoriate`, I'd be fine with any of the serious ones. 

```yaml
delete: foo.txt
drop: foo.txt
exclude: foo.txt
omit: foo.txt
purge: foo.txt
rm: foo.txt
```

**Comment by oconnor663** [comment]
Another thing I'm thinking about is execution order. We don't document this, but rule fields execute in the order given [here](https://github.com/buildinspace/peru#rules):

1. copy
2. move
3. pick
4. executable
5. export

A new `drop` field (or whatever) would probably go either right before or right after `pick`. Do we care which it is? I'm having trouble thinking of a case where the order matters. (Remember that both fields support globs.)

It's also possible to put it after `export`, and we considered that with `pick` to, but I think we want to stick with before for consistency with `copy`/`move`. That way all the path arguments in `peru.yaml` are full paths.

**Comment by MatthewEppelsheimer** [comment]
When are you likely to use both? In most cases, I will either want one or
the other. If I want to just have a few things (excluding most), I will
`pick` them, and won't need to `drop` anything else (that's implicit). If I
want everything except for a few things, I will `drop` the things to
exclude, and `pick`-ing everything else is implicit.

In some cases, using them in combination could be nice: Writing overly
broad `pick` paths, then using `drop` on a few more explicit things the
user doesn't actually want, could be both faster for the user to write, and
also much cleaner to read than a list of more explicit `pick` paths. To
support this use case, `drop` needs to come after `pick`.

Conversely, the reverse order doesn't make sense: explicitly excluding some
things, before we've included anything, and then including some things, is
nonsensical, or at least inefficient, because that's the same as just
including some things. So executing `drop` before `pick` doesn't seem to
make sense. (This may be overly simplistic for someone something much more
complicated with rules, but I can't think of what that would be.)

It seems best to execute `drop` (or whatever) _after_ `pick`.

This was fun to think through.


On Wed, Apr 20, 2016 at 10:07 AM, oconnor663 <notifications@github.com>
wrote:

> Another thing I'm thinking about is execution order. We don't document
> this, but rule fields execute in the order given here
> <https://github.com/buildinspace/peru#rules>:
>
>    1. copy
>    2. move
>    3. pick
>    4. executable
>    5. export
>
> A new drop field (or whatever) would probably go either right before or
> right after pick. Do we care which it is? I'm having trouble thinking of
> a case where the order matters. (Remember that both fields support globs.)
>
> It's also possible to put it after export, and we considered that with
> pick to, but I think we want to stick with before for consistency with
> copy/move. That way all the path arguments in peru.yaml are full paths.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/buildinspace/peru/issues/150#issuecomment-212517449>
>



-- 

Matthew Eppelsheimer
Rocket Lift

(503) 928-3478
matthew@rocketlift.com

**Comment by oconnor663** [comment]
I imagine the use case where you might want both is like you described, where one of the two is doing broad strokes like globs or whole directories, and the other is just listing specific exceptions. Imagine we have this directory:

```
.
├── a.txt
├── b.txt
├── c.txt
└── d.sh
```

We want "all the .txt files except b.txt", so we write these fields:

```yaml
- pick: "*.txt"
- drop: "b.txt"
```

I believe that will work regardless of the order in which drop and pick are executed. If drop goes first, then pick's glob won't see `b.txt` at all. If pick goes first, the glob will match it, but then it'll get removed afterwards by drop.

I _think_ this is true of all cases? Haven't convinced myself yet. @olson-sean-k?

**Comment by oconnor663** [comment]
Here's [the code for all this](https://github.com/buildinspace/peru/blob/master/peru/rule.py#L40-L52).

**Comment by oconnor663** [comment]
Another interaction to consider is that it's an error for any of these globs/literals to fail to match at least one file. So say I have the following fields:

```yaml
drop: foo.json
pick: *.txt
```

I this case the order of execution actually *does* matter. If drop goes first, the rule will succeed. If pick goes first the rule will fail, because drop will match no files, and the user will see `"foo.json" didn't match any files.` But then we could construct an example with the opposite behavior:

```yaml:
drop: *.json
pick: foo.json
```

That one succeeds if pick goes first but fails if drop goes first. So we have (at least) two different categories of user mistakes we could worry about here:

1. dropping files that are implicitly dropped by `pick`
2. picking files that are explicitly dropped by `drop`

If we're only going to have one of those be an actual error -- since I don't think I want to add special error-discovering code besides the current "empty glob" check -- I think case 2 is much more important to enforce. In case 1 peru you've sort of told peru to do the same thing twice, so your rule isn't very well-formed, but you're going to get the result you wanted anyway. But in case 2 your rule contradicts itself, and you're probably going to get confused. An error would be nice here.

**Comment by MatthewEppelsheimer** [comment]
> I believe that will work regardless of the order in which drop and pick are executed. If drop goes first, then pick's glob won't see b.txt at all. (The model here is that each step operates on the tree produced by the previous step, and never actually sees the original tree.) If pick goes first, the glob will match it, but then it'll get removed afterwards by drop.

That is pretty slick. And your description of why `drop` should come before `pick` makes sense given the above.

Another angle then is what the user would expect to happen. Whatever you go with, this operate-on-a-tree nature, and the execution order, should probably be well documented, because it will work differently from (and I think better than) the way I was imagining.

I like the `drop` for the token more and more, btw.

**Comment by oconnor663** [comment]
I just put up a PR for this. @MatthewEppelsheimer and @olson-sean-k, do you guys have time to take it for a spin?

**Comment by oconnor663** [comment]
I went with `delete` in the end because it seemed like the clearest possible name.

**Comment by oconnor663** [comment]
> Whatever you go with, this operate-on-a-tree nature, and the execution order, should probably be well documented, because it will work differently from (and I think better than) the way I was imagining.

By the way, what other semantics were you imagining? I'd like to know so that I can make sure the documentation clarifies that.

**Comment by MatthewEppelsheimer** [comment]
> By the way, what other semantics were you imagining? I'd like to know so that I can make sure the documentation clarifies that.

I had imagined the two operations would be completely separate, without one bearing on the other. This would mean that the order of execution would mean one would override the other. Instead I think of the tree operation as more of a union of the two … if that makes sense.

/me needs more ☕

**Comment by MatthewEppelsheimer** [comment]
> do you guys have time to try it out?

I did some testing, and it's working great. 

In case you're interested, here is my WordPress core-wrangling test case, with module declaration omitted:

```yaml
imports:
  wp-core|skip-default-wp-themes|skip-default-wp-plugins: .

rule skip-default-wp-themes:
  delete: wp-content/themes/twenty*

rule skip-default-wp-plugins:
  delete: wp-content/plugins/hello.php
```

All of WordPress core is synchronized, except for default themes and the Hello Dolly plugin, as intended.

On the subject of documentation, the new documentation for `delete` says it can accept "a list of files and directories", but I couldn't figure out how to format that list, to combine these into one rule. Comma- and space-separated paths don't seem to work on a single line, and two discrete `delete` lines in one rule result in the first one being ignored.

**Comment by oconnor663** [comment]
Ugh, I hate that YAML silently drops duplicate lines. There's a [very hacky warning](https://github.com/buildinspace/peru/blob/master/peru/parser.py#L209) in place to try to detect cases where that's happening though. Did it trigger for you when you duplicated the `delete` field?

As happy as I am to see someone using a multi-rule pipe, you're right that you don't need it in this case. What you need is to give `delete` a YAML list of entries. So either this:

```yaml
rule skip-default-wp-themes:
  delete:
    - wp-content/themes/twenty*
    - wp-content/plugins/hello.php
```

or this

```yaml
rule skip-default-wp-themes:
  delete: ["wp-content/themes/twenty*", "wp-content/plugins/hello.php"]
```

I don't think the quotes are strictly necessary in that second example, but YAML parsing starts to get really dicey as soon as `*` and other interesting syntax gets involved.

I haven't documented this yet, mostly because I felt like the README was getting too long, and I didn't want to weigh it down with too many advanced use cases. I think what I should do now is split off an Advanced document, cram all this stuff in there, and just link to it from the main. We could move a ton of stuff under the fold like that.

**Comment by MatthewEppelsheimer** [comment]
> There's a very hacky warning in place to try to detect cases where that's happening though. Did it trigger for you when you duplicated the delete field?

Yes it did.

Thanks for the assist with the YAML syntax. I'll give that a shot on a break later today.

I'd be happy to assist with the README refactoring — if you don't mind it happening slowly over a few weeks, based on my current schedule. It looks like #110 is a bit stale but a good venue for discussion.

**Comment by olson-sean-k** [comment]
Sorry for chiming in so late. I do recall discussing this in the past, and I'm glad it's being added now. Thanks for putting together the PR.

> But in case 2 your rule contradicts itself, and you're probably going to get confused. An error would be nice here.

Agreed, I think dropping files should definitely happen first.

I'm going to bikeshed a bit and suggest we call this `exclude`, `ignore`, or `drop`. I think these are all more natural antonyms of `pick`. `delete` (reminiscent of `rm`) suggests to me something destructive and irreversible. I think `ignore` or `exclude` are things I'd be likely to search for as well, though not so much for `drop`. (To be fair, `pick` isn't too terribly obvious either. Sometimes you just have to read the manual.)

**Comment by oconnor663** [comment]
I went with `delete` because I thought it was the obvious choice for new users. But if it's not so obvious, then maybe we should just go with `drop` since we all like it :)

**Comment by oconnor663** [comment]
I renamed everything to `drop` [on my branch](https://github.com/buildinspace/peru/pull/152/commits/944ffad3d63b0d9b9daa8cd039ae7b4a8bd17996).

**Comment by oconnor663** [comment]
Will land tonight if there are no objections. By the way, instead of pushing this out as part of version 0.5, I'm thinking about making it 1.0. Can start another issue to coordinate that if anyone has thoughts.

**Comment by MatthewEppelsheimer** [comment]
No objections to landing with `drop` (or anything else).

**Comment by olson-sean-k** [comment]
`drop` looks good to me!

A bit unrelated, but before bumping to 1.0, maybe we can review open issues and discuss?

**Comment by oconnor663** [comment]
Landed! @olson-sean-k cleaned out the [1.0 milestone](https://github.com/buildinspace/peru/milestones/one%20point%20oh!) this evening, and what's left is pretty small. I think I'd like to make `drop` part of that release, @MatthewEppelsheimer, if you don't mind waiting a week or two / using master until then.

**Comment by oconnor663** [comment]
@MatthewEppelsheimer we tagged the 1.0 release yesterday, so this is up on PyPI now.
